### DISCUSSION TITLE
Can I programmatically open a directory when invoking `opensshremotes.openEmptyWindow`

### DISCUSSION THREAD

**Discussion by jasongonzales23** [microsoft/vscode-discussions#58] [Extension Development QnA] [answered]
Our extension connects users to remote environments, but our users are rightly complaining of a bad experience. This is because every time they connect to a remote, there are rituals they have to repeat. Among them are clicking "continue", agreeing that they trust the thing they are about to open and then finally navigating to the same folder every time. We put customer code into a directory called "projects." The thing is we can navigate to the folder for them if they come from the web. We simply pass that in as part of a vscode URL. That said, I don't see how we can do the same thing from within an extension. If this is possible via VScode URL it should be allowed from an extension IMO.

Furthermore, (and I realize this is a separate ask) it would be nice to give users the ability to "never do this again" with the other rituals as well. It's a bad experience for them.

**Comment by alefragnani** [comment, marked as answer]
Hi @jasongonzales23 ,

First, about the `trust` message. If you are talking about the `Workspace Trust` feature, I suppose the Remote extension has no _special treatment_ on that matter. So, it simply inherits the VS Code setting. If that the case, and the user is annoyed about that AND it does trust on remotes, he/she could simply disable the `security.workspace.trust.enabled` setting. I'm not aware of any _Remote specific trust setting_, but I guess it would be a good feature request for VS Code.

About the API, It seems `opensshremotes.openEmptyWindow` does indeed support parameters, but not exactly what you need. Based on this issue (https://github.com/microsoft/vscode-remote-release/issues/3552) you are able to pass some connection parameters, but not the folder you would like to navigate. Unfortunately, the Remote extensions are not open source, neither documented (for extension authors), so it will be hard to identify if the API has evolved.

I'm not sure the developers of that extension are aware of this Discussions repo (they weren't aware of the Slack channel), so I suggest you to ask your question on their repo as well (https://github.com/microsoft/vscode-remote-release/issues/).

**Reply by alefragnani** [reply]
Just a  question, why do you use `opensshremotes.openEmptyWindow`, if the user needs to _navigate to the same folder every time_?

If you have to open the same folder every time, why don't you just use the `vscode.openFolder` command instead, with the remote path as parameter? VS Code will properly start the folder in SSH.

My [Project Manager](https://marketplace.visualstudio.com/items?itemName=alefragnani.project-manager) extension does that, and opens SSH, WSL, Containers, Codespaces, just fine. 

I don't have a SSH project available, but a WSL path is stored like this: `vscode-remote://wsl+ubuntu/home/sources/my-specific-project`

Hope this helps

**Reply by jasongonzales23** [reply]
Hey thanks for this, I hadn't thought of this approach!

**Reply by alefragnani** [reply]
You’re welcome. Let us know if it works for you.

**Reply by jasongonzales23** [reply]
Struggling with this a bit this morning. Could be due to my unfamiliarity with how to work with VSCode, but I believe in my defense the docs are pretty unclear. Here is how I'm trying to use your suggestion and here is the error I get:

```
  const uri = `vscode-remote://ssh-remote+${machineAddress}/home/folder/sub-folder`;
  vscode.commands.executeCommand("vscode.openFolder", uri, {forceNewWindow: true});
```

Error:
```
stack trace: Error: argument does not match one of these constraints: arg instanceof constraint, arg.constructor === constraint, nor constraint(arg) === true
```
That error is not at all helpful to me but I suspect maybe it doesn't like the uri as a string? I'll renew my gripe about VSCode extension docs here. How am I to know how to form the arguments and what types this command expects if they aren't in code hints or the docs. I suspect you will tell me to read source code, but I find that answer frustrating for reasons I think should be obvious.

**Reply by jasongonzales23** [reply]
Hm ok, after trial and error, I got progress but not success. Here's the code:
```
  const uri = vscode.Uri.parse(
    `vscode-remote://ssh-remote+${machineAddress}/home/folder/sub-folder`
  );

  vscode.commands.executeCommand("vscode.openFolder", uri, {
    forceNewWindow: true,
  });
```

However, I get "denied access public key" and I think it's because I need to pass along a username. Any idea how I do that here?

**Reply by alefragnani** [reply]
The `uri` param is not a `string`, but a `Uri` object. 

Try this:

```
const uri = vscode.Uri.parse(`vscode-remote://ssh-remote+${machineAddress}/home/folder/sub-folder`)
vscode.commands.executeCommand("vscode.openFolder", uri, {forceNewWindow: true});
```

If the `path` is not _remote_ you could use `vscode.Uri.file` instead

Hope this helps

**Reply by jasongonzales23** [reply]
Thanks @alefragnani ! I tried exactly what you suggest above and got the "access denied public key" error. What's even stranger is that we have a flow that goes from the web where we simply offer a link on the web with the following:

```
 const remoteVsCodeUrl = `vscode://vscode-remote/ssh-remote+${machineAddress}/home/folder/sub-foilder`;
```

Click on that link from the web and VSCode behaves as we'd expect.

**Reply by alefragnani** [reply]
Not sure why this error message, unfortunately. I have no SSH available to try out right now.
But it's really weird that it works from the web, but not on desktop 😕

**Reply by jasongonzales23** [reply]
Hey @alefragnani I figured out the issue and thanks so much for the help. It may be a quirk of our implementation but I needed to add `username@machineAddress` for this all to work.

**Reply by alefragnani** [reply]
Good to know it is working!
